### PR TITLE
GsnDevProvider

### PR DESCRIPTION
**Pull request by drortirosh**
dev provider: brings up a Relay server inside the provider, to simplify local testing of contracts.

### CODE REVIEW

**Comment by forshtat** [comment]
please do not merge before TS rewrite PR!

File: src/relayclient/ContractInteractor.ts [hunk 1]
```diff
@@ -28,8 +28,8 @@ let IRelayHubContract: Contract<IRelayHubInstance>
 let IForwarderContract: Contract<ITrustedForwarderInstance>
 
 export default class ContractInteractor {
-  private readonly web3: Web3
-  private readonly provider: provider
+  readonly web3: Web3
+  readonly provider: provider
```
Comment (forshtat) [inline src/relayclient/ContractInteractor.ts:32]: I don't think this is a good idea. Better create an API for the methods you want from the contracts interactor

File: src/relayclient/DevRelayClient.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    const stakeManagerAddress = await hub.getStakeManager()
+
+    // flag early, so only the first call will try to bring up a relay
+    // (TODO: other calls should still wait for the relay to start)
+    this.serverStarted = true
+
+    // const {
+    //   relayOwner, relayWorkdir, relayHub, relayListenPort,
+    //   baseRelayFee, pctRelayFee, gasPriceFactor, devMode
+    // } = this.config
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:196]: \-

File: src/relayclient/DevRelayClient.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    // @ts-ignore
+    return [this.devRelay]
+  }
+
+  // noinspection JSUnusedGlobalSymbols
+  refresh (): void { // ts-ignore
+  }
+
+  devRelay?: RelayRegisteredEventInfo
+}
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:112]: Let's extract an interface, this is ugly to force one class into another, and loses all type protection. Also, https://github.com/opengsn/gsn/pull/324 introduces some changes to this interface.

File: src/relayclient/DevRelayClient.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+  /**
+     * Options include standard transaction params: from,to, gas_price, gas_limit
+     * relay-specific params:
+     *  pctRelayFee (override config.pctRelayFee)
+     *  validateCanRelay - client calls canRelay before calling the relay the first time (defaults to true)
+     *  paymaster - the contract that is compensating the relay for the gas (defaults to transaction destination 'to')
+     * can also override default relayUrl, relayFee
+     * return value is the same as from sendTransaction
+     */
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:154]: Is this comment copied from somewhere? It does not sound up-to-date.

File: src/relayclient/DevRelayClient.ts [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  async _initializeRelay (): Promise<void> {
+    if (this.serverStarted) {
+      return
+    }
+
+    const hub = await this.contractInteractor._createRelayHub(this.config.relayHubAddress)
+    const stakeManagerAddress = await hub.getStakeManager()
+
+    // flag early, so only the first call will try to bring up a relay
+    // (TODO: other calls should still wait for the relay to start)
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:190]: I don't think there is a way to have 'other calls' unless someone creates multiple promises of `relayTransaction` without awaiting on them, which is something we generally never do.

File: src/relayclient/DevRelayClient.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    this.httpServer = httpServer
+
+    // @ts-ignore
+    const IStakeManagerContract: Contract<IStakeManagerInstance> = TruffleContract({
+      contractName: 'IStakeManager',
+      abi: stakeManagerAbi
+    })
+    IStakeManagerContract.setProvider(this.contractInteractor.provider, undefined)
+
+    const stakeManager = await IStakeManagerContract.at(stakeManagerAddress)
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:214]: Created this in Contract Interactor, PR#324

File: src/relayclient/DevRelayClient.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+    const estim = await stakeManager.contract.methods.stakeForAddress(relayServer.address, weekInSec)
+      .estimateGas({ value: 1e18 })
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:217]: Why is this necessary? Cannot it estimate on its own?

File: src/relayclient/DevRelayClient.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    // owner: relayOwner,
+    hubAddress: relayHub,
+    url: relayUrl,
+    baseRelayFee: devConfig.baseRelayFee,
+    pctRelayFee: devConfig.pctRelayFee,
+    gasPriceFactor: devConfig.gasPriceFactor,
+    devMode: devConfig.devMode
+  })
+  relayServer.on('error', (e: any) => {
+    // console.error('ERR:', e.message)
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:75]: Do emit if verbose, please. This commented-out outputs do shoot you in the back sometimes.

File: src/relayclient/DevRelayClient.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    provider: HttpProvider,
+    devConfig: Partial<DevGSNConfig>,
+    overrideDependencies?: Partial<GSNDependencies>
+  ) {
+    super(provider, devConfig,
+      {
+        ...overrideDependencies,
+        knownRelaysManager: new DevKnownRelays() as KnownRelaysManager
+      })
+    this.devConfig = this.config as DevGSNConfig
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:143]: You force-cast `Partial<DevGSNConfig>` to `DevGSNConfig`, that would mean bailing on TypeScript's null-safety. Not so much here, as almost every field in DevGSNConfig is optional anyways. 
This is complicating the understanding of what is going on. May I suggest `DevGSNConfig` to have all fields initialised, like `GSNConfig` does, and create a 'config resolver' method similar to `configureGSN`? 
I think this approach makes it way easier to follow the configuration flow.

File: src/relayclient/DevRelayClient.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+          pctRelayFee: '0',
+          baseRelayFee: '0',
+          // @ts-ignore
+          relayUrl: relayServer.url
+        }
+      }
+    })
+
+    const devRelays = this.knownRelaysManager as DevKnownRelays
+    devRelays.devRelay = relayInfo
```
Comment (forshtat) [inline src/relayclient/DevRelayClient.ts:246]: It is not a pay-per-word message, and I bet your screen can handle more then 480 characters in a row :-)

File: src/relayclient/GsnDevProvider.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import RelayProvider from './RelayProvider'
+import { DevRelayClient, DevGSNConfig } from './DevRelayClient'
+import { HttpProvider } from 'web3-core'
+import { configureGSN, GSNDependencies } from './GSNConfigurator'
+
+export default class GsnDevProvider extends RelayProvider {
+  devRelayClient: DevRelayClient
```
Comment (forshtat) [inline src/relayclient/GsnDevProvider.ts:7]: is there a reason it should not be `private readonly`?

File: src/relayclient/RelayClient.ts [hunk 12]
```diff
@@ -39,11 +39,11 @@ export interface RelayingResult {
...
 export default class RelayClient {
-  private readonly config: GSNConfig
+  readonly config: GSNConfig
   private readonly httpClient: HttpClient
-  private readonly contractInteractor: ContractInteractor
-  private readonly knownRelaysManager: KnownRelaysManager
-  private readonly accountManager: AccountManager
+  readonly contractInteractor: ContractInteractor
+  readonly knownRelaysManager: KnownRelaysManager
+  readonly accountManager: AccountManager
```
Comment (forshtat) [inline src/relayclient/RelayClient.ts:46]: no

File: test/relayclient/RelayProvider.test.ts [hunk 13]
```diff
@@ -161,7 +161,7 @@ contract('RelayProvider', function (accounts) {
     let notRelayedTxReceipt: BaseTransactionReceipt
     let misbehavingPaymaster: TestPaymasterConfigurableMisbehaviorInstance
     let relayRequest: RelayRequest
-    // It is not strictly necessary to make this test against actual tx receipt, but I prefer to do it anyway
+    // It is not stirctly necessary to make this test against actual tx receipt, but I prefer to do it anyway
```
Comment (forshtat) [inline test/relayclient/RelayProvider.test.ts:164]: I prefer `strictly` to `stirctly`, don't you?

File: src/relayclient/DevRelayClient.ts (hunk 3 above)
Comment (drortirosh) [reply src/relayclient/DevRelayClient.ts]: done
